integrations/datadog/client.py

```diff
@@ -233,6 +296,373 @@ async def get_slo_history(
+            ]
+
+            if queries:
+                matching_queries[widget["id"]] = {
+                    "widget_id": widget["id"],
+                    "widget_title": widget["definition"].get("title", "Untitled"),
+                    "queries": queries,
+                }
+
+        return list(matching_queries.values())
+
+    async def get_single_dashboard(self, dashboard_id: str) -> dict[str, Any] | None:
+        if not dashboard_id:
+            return None
+        url = f"{self.api_url}/api/v1/dashboard/{dashboard_id}"
+        return await self._send_api_request(url)
+
+    async def enrich_dashboard_url(self, dashboard: dict[str, Any]) -> dict[str, Any]:
+        """
+        Enriches a dashboard item with the full Datadog web URL.
+        """
+        dashboard["url"] = f"{self.datadog_web_url}{dashboard['url']}"
+        return dashboard
+
+    async def get_dashboards(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        page = 0
+        page_size = MAX_PAGE_SIZE
+
+        while True:
+            url = f"{self.api_url}/api/v1/dashboard"
+            result = await self._send_api_request(
+                url,
+                params={
+                    "start": page,
+                    "count": page_size,
+                    "filter[shared]": "false",
+                    "filter[deleted]": "false",
+                },
+            )
+            dashboards = result.get("dashboards")
+            if not dashboards:
+                break
+
+            # Enrich items concurrently using asyncio.gather
+            dashboards = await asyncio.gather(
+                *[self.enrich_dashboard_url(dashboard) for dashboard in dashboards]
+            )
+
+            yield dashboards
+            page += 1
+
+    async def add_dashboard_id_to_widget(
+        self, widget: dict[str, Any], dashboard_id: str
+    ) -> dict[str, Any]:
+        """
+        Adds the dashboard ID to a widget dictionary.
+        """
+        widget["dashboard_id"] = dashboard_id
+        return widget
+
+    async def get_dashboard_metrics(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        """
+        Asynchronously fetches and yields lists of widgets (metrics) from multiple Datadog dashboards.
+
+        This method retrieves dashboards in batches, processes them concurrently to extract widgets (metrics),
+        and yields each list of widgets as they become available.
+
+        Yields:
+            Lists of dictionaries, where each dictionary represents a widget (metric) from a Datadog dashboard.
+            The structure of each widget dictionary depends on the Datadog API response.
+        """
+        async for dashboards in self.get_dashboards():
+            metrics = await process_in_queue(
+                [dashboard["id"] for dashboard in dashboards],
+                self.get_single_dashboard,
+                concurrency=MAX_CONCURRENT_REQUESTS,
+            )
+
+            for metric in metrics:
```

Comment:
I am not familiar with datadog dashboard but does the widget contain the metrics? From the docs, I only see widgets instead of metrics

Reply:
Every widget contains metrics.